test-tools/src/main/java/org/terracotta/utilities/test/logging/ConsoleAppenderCapture.java

```diff
@@ -0,0 +1,235 @@
+          String target = ((ConsoleAppender<?>)iLoggingEventAppender).getTarget();
+          ListAppender<ILoggingEvent> appender = appenderMap.computeIfAbsent(target, t -> {
+            ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+            listAppender.setContext(context);
+
+            listAppender.addFilter(new AbstractMatcherFilter<ILoggingEvent>() {
+              @Override
+              public FilterReply decide(ILoggingEvent iLoggingEvent) {
+                if (iLoggingEvent.getMDCPropertyMap().getOrDefault(mdcKey, "").equals(ConsoleAppenderCapture.this.testId)) {
+                  return FilterReply.ACCEPT;
+                } else {
+                  return FilterReply.DENY;      // Filter only applies to 'this' Appender -- can stop processing here
+                }
+              }
+            });
+
+            listAppender.start();
+            return listAppender;
+          });
+          logger.addAppender(appender);
+          Runnable r = removeAppenders;
+          removeAppenders = () -> {
+            r.run();
+            logger.detachAppender(appender);
+            appender.stop();
+          };
+          break;
+        }
+      }
+    }
+
+    this.removeAppenders = removeAppenders;
+  }
+
+  /**
+   * Gets a reference to the {@code List}s into which log entries are captured.
+   *
+   * @return the map of logging target to log entry list reference; the key of the
+   *      map is the {@link ConsoleAppender#getTarget()} value for which
+   *      {@link Target#targetName()} value may be used
+   * @see Target
+   */
+  public Map<String, List<ILoggingEvent>> getLogs() {
+    return appenderMap.entrySet().stream()
+        .collect(toMap(Map.Entry::getKey, e -> e.getValue().list));
+  }
+
+  /**
+   * Gets the events, in formatted string form, logged to the specified target.
+   * @param target the target for which log messages are to be returned
+   * @return the list of messages logged to {@code target}; may be empty
+   */
+  public List<String> getMessages(Target target) {
+    requireNonNull(target, "target");
+    return getLogs().getOrDefault(target.targetName(), Collections.emptyList())
+        .stream().map(ILoggingEvent::getFormattedMessage).collect(toList());
+  }
+
+  /**
+   * Gets the events, in formatted string form, as a single string with events
+   * separated by {@link System#lineSeparator()}.
+   * @param target the target for which log messages are to be returned
+   * @return a string containing all messages logged to {@code target}; may be empty
+   */
+  public String getMessagesAsString(Target target) {
+    requireNonNull(target, "target");
+    return getLogs().getOrDefault(target.targetName(), Collections.emptyList())
+        .stream().map(ILoggingEvent::getFormattedMessage).collect(joining(System.lineSeparator()));
+  }
+
+  /**
+   * Detaches and stops the capturing appenders.
+   */
+  @Override
+  public void close() {
+    removeAppenders.run();
```

Comment:
`removeAppenders` could end up closing IO things right? Should there be some exception handling/finallyl logic going on here?

Reply:
The only appender involved here is the `ListAppender` -- no I/O.  Even if it were something like the `FileAppender`, any I/O errors raised are captured and presented to the `StatusAppender` that can be observed by a `StatusListener` or directly through the `StatusManager`.  No exception is thrown.